vowpalwabbit/memory_tree.cc

```diff
@@ -0,0 +1,1358 @@
+        ec.weight = 1.f;
+	    ec.l.simple = {route_label, 1., 0.f};
+        base.learn(ec, b.nodes[cn].base_router); //update the router according to the new example.
+        
+        base.predict(ec, b.nodes[cn].base_router);
+        float save_binary_scalar = ec.pred.scalar;
+
+        if(b.oas == false){
+            ec.l.multi = mc;
+            ec.pred.multiclass = save_multi_pred;
+        }
+        else{
+            ec.pred.multilabels = preds;
+            ec.l.multilabels = multilabels;
+        }
+        ec.weight = ec_input_weight;
+
+        return save_binary_scalar;
+    }
+
+    
+    //turn a leaf into an internal node, and create two children
+    //when the number of examples is too big
+    void split_leaf(memory_tree& b, single_learner& base, const uint32_t cn)
+    {
+        //create two children
+        b.nodes[cn].internal = 1; //swith to internal node.
+        uint32_t left_child = (uint32_t)b.nodes.size();
+        b.nodes.push_back(node());
+        b.nodes[left_child].internal = -1;  //left leaf
+        b.nodes[left_child].base_router = (b.routers_used++);
+        uint32_t right_child = (uint32_t)b.nodes.size();
+        b.nodes.push_back(node());  
+        b.nodes[right_child].internal = -1;  //right leaf
+        b.nodes[right_child].base_router = (b.routers_used++); 
+
+        if (b.nodes[cn].depth + 1 > b.max_depth){
+            b.max_depth = b.nodes[cn].depth + 1;
+            cout<<"depth "<<b.max_depth<<endl;
+        }
+
+        b.nodes[cn].left = left_child;
+        b.nodes[cn].right = right_child;
+        b.nodes[left_child].parent = cn;
+        b.nodes[right_child].parent = cn;
+        b.nodes[left_child].depth = b.nodes[cn].depth + 1;
+        b.nodes[right_child].depth = b.nodes[cn].depth + 1;
+
+        if (b.nodes[left_child].depth > b.max_depth)
+            b.max_depth = b.nodes[left_child].depth;
+
+        //rout the examples stored in the node to the left and right
+        for(size_t ec_id = 0; ec_id < b.nodes[cn].examples_index.size(); ec_id++) //scan all examples stored in the cn
+        {
+            uint32_t ec_pos = b.nodes[cn].examples_index[ec_id];
+            MULTICLASS::label_t mc;
+            uint32_t save_multi_pred;
+            MULTILABEL::labels multilabels;
+            MULTILABEL::labels preds;
+            if (b.oas == false){
+                mc = b.examples[ec_pos]->l.multi;
+                save_multi_pred = b.examples[ec_pos]->pred.multiclass;
+            }
+            else{
+                multilabels = b.examples[ec_pos]->l.multilabels;
+                preds = b.examples[ec_pos]->pred.multilabels;
+            }
+
+            b.examples[ec_pos]->l.simple = {1.f, 1.f, 0.f};
+            base.predict(*b.examples[ec_pos], b.nodes[cn].base_router); //re-predict
+            float scalar = b.examples[ec_pos]->pred.scalar;
+            if (scalar < 0)
+            {
+                b.nodes[left_child].examples_index.push_back(ec_pos);
```

Comment:
Shouldn't the outcome of train_node be determining whether it's add to the left or right child?

Reply:
here we are splitting a leaf whose number of stored examples exceeds limit. We use the current learner's prediction in this node to split it.

Reply:
Ah, we are training a splitter for the leaf with each example incident rather than training a batch when it is needed?